modules/gdnative/include/pluginscript/godot_pluginscript.h

```diff
@@ -0,0 +1,171 @@
+	godot_bool (*validate)(godot_pluginscript_language_data *p_data, const godot_string *p_script, int *r_line_error, int *r_col_error, godot_string *r_test_error, const godot_string *p_path, godot_pool_string_array *r_functions);
+	int (*find_function)(godot_pluginscript_language_data *p_data, const godot_string *p_function, const godot_string *p_code); // Can be NULL
+	godot_string (*make_function)(godot_pluginscript_language_data *p_data, const godot_string *p_class, const godot_string *p_name, const godot_pool_string_array *p_args);
+	godot_error (*complete_code)(godot_pluginscript_language_data *p_data, const godot_string *p_code, const godot_string *p_base_path, godot_object *p_owner, godot_array *r_options, godot_bool *r_force, godot_string *r_call_hint);
+	void (*auto_indent_code)(godot_pluginscript_language_data *p_data, godot_string *p_code, int p_from_line, int p_to_line);
+
+	void (*add_global_constant)(godot_pluginscript_language_data *p_data, const godot_string *p_variable, const godot_variant *p_value);
+	godot_string (*debug_get_error)(godot_pluginscript_language_data *p_data);
+	int (*debug_get_stack_level_count)(godot_pluginscript_language_data *p_data);
+	int (*debug_get_stack_level_line)(godot_pluginscript_language_data *p_data, int p_level);
+	godot_string (*debug_get_stack_level_function)(godot_pluginscript_language_data *p_data, int p_level);
+	godot_string (*debug_get_stack_level_source)(godot_pluginscript_language_data *p_data, int p_level);
+	void (*debug_get_stack_level_locals)(godot_pluginscript_language_data *p_data, int p_level, godot_pool_string_array *p_locals, godot_array *p_values, int p_max_subitems, int p_max_depth);
+	void (*debug_get_stack_level_members)(godot_pluginscript_language_data *p_data, int p_level, godot_pool_string_array *p_members, godot_array *p_values, int p_max_subitems, int p_max_depth);
+	void (*debug_get_globals)(godot_pluginscript_language_data *p_data, godot_pool_string_array *p_locals, godot_array *p_values, int p_max_subitems, int p_max_depth);
+	godot_string (*debug_parse_stack_level_expression)(godot_pluginscript_language_data *p_data, int p_level, const godot_string *p_expression, int p_max_subitems, int p_max_depth);
+
+	// TODO: could this stuff be moved to the godot_pluginscript_language_desc ?
+	void (*get_public_functions)(godot_pluginscript_language_data *p_data, godot_array *r_functions);
+	void (*get_public_constants)(godot_pluginscript_language_data *p_data, godot_dictionary *r_constants);
+
+	void (*profiling_start)(godot_pluginscript_language_data *p_data);
+	void (*profiling_stop)(godot_pluginscript_language_data *p_data);
+	// ProfilingInfo dictionary format
+	// {
+	//   <signature:str>: {
+	//     call_count: <int>
+	//     total_time: <int:microseconds>
+	//     self_time: <int:microseconds>
+	//   }
+	// }
+	int (*profiling_get_accumulated_data)(godot_pluginscript_language_data *p_data, godot_dictionary *r_info, int p_info_max);
```

Comment:
I'm not sure about how `ProfilingInfo` is passed here:

In the end pluginscript will fill an a C array of `ProfilingInfo` structs.

Currently the language will fill a godot dict, with function signature as key and a dict of ints as value.

So Pluginscript iterate over the dict to fill the final C array. Meaning:
- creating the dict with function signature as key takes time for each insertion for no benefit (given the dict is only iterated)
- creating all the subdicts is costly
- the iteration & recopy of the dicts into the `ProfilingInfo` class instances takes time

Given `profiling_get_frame_data` is called for each frame during debug, I think we should try to lower the overhead, the possibilities I see:
1) replace the godot dict returned by the function by an godot array
2) create a new C struct equivalent `ProfilingInfo` and allocate an array of them to be filled by the `profiling_get_frame_data`
3) pass the actual C array of `ProfilingInfo` class instances to `profiling_get_frame_data`, and allow it to be filled with a custom function:
```c
void godot_pluginscript_fill_profiling_entry(void *profiling_array, int index, godot_string_name *signature, godot_int call_count, godot_int total_time, godot_int self_time) {
  ProfilingInfo *array = profiling_array;
  array[index].signature = StringName(*(StringName*)signature);
  array[index].call_count = call_count;
  array[index].total_time = total_time;
  array[index].self_time = self_time;
}
```

@karroffel any idea on this ?

Reply:
I like 2.

Reply:
done.